### PR TITLE
implement callback functions

### LINKED ISSUE #78: Post prediction hook that has access to features, model, prediction for e.g. WhyLogs, Revela integration
Use case:
As a data scientist, I want to define a custom function hook after the `model.predictor` function has been called so that I can track metrics and aggregates about the features and predictions, something like:

```python
def my_custom_callback(model: ModelType, features: FeatureType, prediction: PredictionType):
    # do something
    ...

# alternatively, something more flexible would be:
def my_custom_callback(inputs: PredictCallbackInputs):
    # do something
    inputs.model
    inputs.features
    inputs.prediction
    inputs.* # whatever else might be needed in the future
    ...


@model.predictor(post_callback=my_custom_callback)
def predictor(model: ModelType, features: FeatureType) -> PredictionType:
    predictions = ...
    return predictions
```

### PR DESCRIPTION
**Pull request by zevisert**
Closes: #78 
Related: #134 

## Changes proposed (Describe Changes)

This PR opens the door to custom post-prediction callbacks. Background: Revela is building an API focused ML monitoring system. We already can support UnionML predictions by adding our logging code to the body of a `@model.predictor` decorated function - but from #78, it seems like it would be preferrable to use a different flyte task to handle post-prediction work.

A prediction callback function receives 3 arguments and returns nothing:  
1. The model instance that made the predictions
2. The input data used to invoke the model
3. The model's outputs

Example:
```py

def some_callback(estimator: LogisticRegression, features: pd.DataFrame, predictions: List[float]):
    pass

@model.predictor(callbacks=[some_callback])
def predictor(estimator: LogisticRegression, features: pd.DataFrame) -> List[float]:
   ...
```

With this implementation, all callbacks must match their parameter type annotations to those defined by the `predictor`. `TypeError`s will be raised otherwise.

## Check List (Check all the applicable boxes) 

- [x] My code follows the code style of this project.
- [x] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] This PR does not contain plagiarized content.
- [x] The title of my pull request is a short description of the requested changes.

## Screenshots

<details>
<summary>Screenshots from an earlier iteration of this change showing callback functions executing as separate tasks on the flyte console</summary>

For workflows with features:
image

For workflows without features:
image

</details>

## Note to Reviewers

- [ ] I need to add integration tests still, just testing the waters on this feature first.
- ~~I **don't** know flyte at all really, so I may be doing things oddly. I initially tried `wf.add_subwf` but couldn't get the input bindings to work correctly -- is a sub workflow a better solution than just conditionally adding tasks?~~
  - For all of my interactions with flyte, assume that I didn't choose the way I did something for any particular reason, call me out on odd usage that you see.
- [x] I left a handful of comments like `# TODO(reveiw,zevisert)` for spots that I had questions that would be best resolved here

## Open in Gitpod

<a href="https://gitpod.io/#https://github.com/unionai-oss/unionml/pull/195"></a>

### CODE REVIEW

File: unionml/model.py [hunk 1]
```diff
@@ -109,6 +109,8 @@ class in sklearn, ``Module`` in pytorch, etc.
         self._train_task = None
         self._predict_task = None
         self._predict_from_features_task = None
+        self.__callbacks: Dict[str, Callable] = {}
```
Comment (cosmicBboy) [inline unionml/model.py:112]: we'll potentially want to add callbacks for the `trainer` and `evaluator`, so let's rename this to `__predict_callbacks`

File: unionml/model.py [hunk 2]
```diff
@@ -109,6 +109,8 @@ class in sklearn, ``Module`` in pytorch, etc.
...
+        self.__callback_tasks: Dict[str, Any] = {}
```
Comment (cosmicBboy) [inline unionml/model.py:113]: callbacks-as-tasks won't be necessary if we invoke them inside the `@inner_task(...)(predict_task)` function, as suggested above.

File: unionml/model.py [hunk 3]
```diff
@@ -119,6 +121,25 @@ class in sklearn, ``Module`` in pytorch, etc.
 
         self._patch_destination_dir: Optional[str] = None
 
+    @property
+    def callbacks(self) -> Dict[str, Callable]:
```
Comment (cosmicBboy) [inline unionml/model.py:125]: rename to `predict_callbacks` to disambiguate from other callbacks

File: unionml/model.py [hunk 4]
```diff
@@ -358,6 +395,10 @@ def predict_workflow(self):
         )
         for output_name, promise in predict_node.outputs.items():
             wf.add_workflow_output(output_name, promise)
+
+        if self.callbacks:
```
Comment (cosmicBboy) [inline unionml/model.py:399]: this won't be needed if callbacks are invoked in `predict_task`

File: unionml/model.py [hunk 5]
```diff
@@ -570,12 +616,130 @@ def predict(
...
+        actual_task = inner_task(
+            unionml_obj=self,
+            input_parameters=OrderedDict([(p.name, p) for p in [model_param, features_param, prediction_param]]),
+            return_annotation=None,
+            task_config={"unionml:variant": callback.__name__},
+        )(callback_from_features_task)
+
+        self.__callback_tasks[task_name] = actual_task
+
+        return actual_task
```
Comment (cosmicBboy) [inline unionml/model.py:742]: this won't be needed if callbacks are invoked in `predict_task`

**Review by cosmicBboy — COMMENTED** [review]
hey @zevisert this is a great start! I think we can vastly simplify the changes here. With UnionML we don't have to create new Flyte tasks for the different callbacks: the callback can essentially wrap the predictor function.

Basically all we need to do is to invoke the callbacks here:
https://github.com/unionai-oss/unionml/blob/main/unionml/model.py#L471

Right after the `self._predictor` is invoked, like:

```python
        def predict_task(model_object, **kwargs):
            parsed_data = self.dataset._parser(kwargs[data_arg_name], **self._dataset.parser_kwargs)
            features = self.dataset._feature_transformer(parsed_data[self._dataset._parser_feature_key])
            predictions = self._predictor(model_object, features)
            for callback in self._predict_callbacks:
                callback(model_object, features, prediction)
            return predictions
```

**Comment by zevisert** [comment]
> With UnionML we don't have to create new Flyte tasks for the different callbacks: the callback can essentially wrap the predictor function.
> 
> Basically all we need to do is to invoke the callbacks here: https://github.com/unionai-oss/unionml/blob/main/unionml/model.py#L471

Sure, updated to the changes here to take the simpler approach. I was trying to keep the callbacks separate for a few reasons (probably, all premature optimization). Here's a few thoughts on it. Correct me if I'm wrong, but placing the callbacks in the same prediction task...
- ... could mean that the flyte task can fail because of a failed callback.
  - I tried to isolate that in the current version with `try: ... except Exception:` so that a callback function can fail without failing the task.
- ... requires the flyte task has to wait for all callbacks to complete before it can complete.
  - Poor parallelism here, but sure, most users will probably only have one or two callbacks. We could run them in parallel on our own with `multiprocessing` or `threading` to help if that's a concern
- ... requires the FastAPI /predict endpoint to wait for callbacks to complete before responding.
  - I was hoping to put callbacks in FastAPI's BackgroundTasks so that the request can complete with the predictions before resolving the callbacks.

File: unionml/type_guards.py [hunk 6]
```diff
@@ -169,6 +169,67 @@ def guard_predictor(predictor: Callable, expected_model_type: Type, expected_dat
         raise TypeError("The 'predictor' function needs a return type annotation.")
 
 
+def guard_callback(callback: Callable, predictor: Callable, expected_model_type: Type, expected_data_type: Type):
```
Comment (cosmicBboy) [inline unionml/type_guards.py:172]:
```suggestion
def guard_prediction_callback(callback: Callable, predictor: Callable, expected_model_type: Type, expected_data_type: Type):
```

File: unionml/model.py [hunk 7]
```diff
@@ -570,12 +619,19 @@ def predict(
...
+            wf = self.predict_workflow()(model_object=self.artifact.model_object, **reader_kwargs)
+        else:
+            wf = self.predict_from_features_workflow()(
+                model_object=self.artifact.model_object,
+                features=self._dataset.get_features(features),
+            )
+
+        # TODO(review,zevisert): Should local predict run callbacks too?
+
+        return wf
```
Comment (cosmicBboy) [inline unionml/model.py:634]: what's the purpose of this change?

re: the TODO question: I think for this PR we can have callbacks invoked everywhere. I'd imagine it would be useful for testing.

We can then think about ways in which we can suppress certain callbacks in different situations, esp. if we need subtleties around whether they should be called in a Flyte execution or how they should be called in a FastAPI app.

**Comment by cosmicBboy** [comment]
Thanks for listing out the concerns, I think it's good to be aware of them.

This initial implementation is the naive one, and if any issues come up with regard to the 3 that you listed (potentially maybe more) then we can tackle those separately. I think there's an incremental approach here based on certain pains that may come up later.

> requires the FastAPI /predict endpoint to wait for callbacks to complete before responding.
I was hoping to put callbacks in FastAPI's BackgroundTasks so that the request can complete with the predictions before resolving the callbacks

This would be my biggest concern... how much time would logging via Revela take? Also wondering how FastAPI's BackgroundTasks would work when deploying a AWS lambda function via [FastAPI + Mangum](https://unionml.readthedocs.io/en/stable/serving_aws_lambda.html)

File: unionml/model.py (hunk 7 above)
Comment (cosmicBboy) [reply unionml/model.py:634]: the output of this function should be the result of invoking `wf` locally, so it doesn't actually return a workflow.

**Review by cosmicBboy — APPROVED** [review]
thanks @zevisert ! 🚀
